### ISSUE TITLE
RazorEngine.Generator for VS 2017

### ISSUE THREAD

**Issue by rohatsu** [zihotki/RazorEngine.Generator.Sample#2]
@zihotki Vasili,

We're big fans of RazorEngine.Generator custom tool for VS 2015. I was wondering if you're planning to update it for VS 2017, or if you don't have the time, perhaps open source the code so we could do it ourselves? We'd love to move over to the new VS but lack of the custom tool this is blocking us...

Many thanks,

piotr & rohatsu team

**Comment by zihotki** [comment]
Hey, I will look into it on these weekend and will try to come up with a solution

**Comment by rohatsu** [comment]
Thanks for following up, one of our guys sat down a while back and moved us over to [RazorGenerator](https://github.com/RazorGenerator), there were some differences in generated code, for example around helpers, but we've managed to work around and are happy to stick with it...

**Comment by benosman** [comment]
@rohatsu - do you happen to have any tips to get RazorEngine templates working with RazorGenerator in VS2017?

**Comment by rohatsu** [comment]
Sure, in our case after changing Custom Tool on our views to RazorGenerator we had to do two more things to get them to compile correctly:
* create razorgenerator.directives file in our templates directory with the below content:
```
Namespace: RohatsuWebApp.Templates GeneratePrettyNames: true Generator: Template
```
* used RawStringFactory in TemplateServiceConfiguration constructor:
```
var config = new RazorEngine.Configuration.TemplateServiceConfiguration
{
     EncodedStringFactory = new RawStringFactory(),
     ...
```
The above was only for non-MVC .cshtml templates, MVC views we already had on RazorGenerator.

**Comment by benosman** [comment]
@rohatsu - Many thanks for sharing that ! those instructions worked for me and now the templates get compiled.

However, I am currently using `ResolvePathTemplateManager` to lookup the template files and having the compiled versions makes no difference, the templates used are still from the file. 

Do you use a trick to get RazorEngine to execute a precompiled template? 

I looked into creating a custom `TemplateManager` for this purpose based on the code in this RazorEngine.Generator.Sample repo, but while it almost works - changes to RazorEngine now mean you can't return a compiled template from the `Resolve()` function. So currently I'm looking at creating a custom `CachingProvider` instead, which seems like it may work.

**Comment by rohatsu** [comment]
Yes we are using both a custom CachingProvider and a custom TemplateManager to pre-load everything at startup, take a look at our [RFRazor.cs](https://github.com/rohatsu/riff/blob/master/RIFF.Core/Templates/RFRazor.cs)

**Comment by benosman** [comment]
@rohatsu - thank you once again! :)

That is pretty much the approach I had taken, and had proof of concept working yesterday. Yours is a lot more polished of course, so I will do some refactoring of mine.
